operators/src/pro/executor/replay.rs

```diff
@@ -0,0 +1,736 @@
+                    idx: 0,
+                    waker: None,
+                },
+            )
+            .is_none());
+        Ok(id)
+    }
+
+    fn remove_receiver(&mut self, id: u32) {
+        if let Some(r) = self.receivers.remove(&id) {
+            let idx = r.idx - self.offset;
+            let mut notify = false;
+            for i in idx..self.buffer.len() {
+                let e = &mut self.buffer[i];
+                e.expected_receivers -= 1;
+                notify |= e.expected_receivers == 0;
+            }
+            if notify {
+                self.notify_senders();
+            }
+        }
+    }
+
+    fn try_recv(&mut self, receiver_id: u32) -> Result<T, ReceiveError> {
+        let c = self.receivers.get_mut(&receiver_id).unwrap();
+        let q_idx = c.idx - self.offset;
+
+        let res = match self.buffer.get_mut(q_idx) {
+            Some(e) => {
+                e.expected_receivers -= 1;
+                Ok(e.value.clone())
+            }
+            None if self.senders.is_empty() => Err(ReceiveError::Closed),
+            None => Err(ReceiveError::NoMessage),
+        };
+
+        match res {
+            Ok(v) => {
+                c.idx += 1;
+                self.notify_senders();
+                Ok(v)
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    fn notify_receivers(&mut self) {
+        for v in self.receivers.values_mut() {
+            if let Some(w) = v.waker.take() {
+                w.wake();
+            }
+        }
+    }
+
+    fn remove_receiver_waker(&mut self, receiver_id: u32) {
+        self.receivers.get_mut(&receiver_id).unwrap().remove_waker();
+    }
+
+    fn update_receiver_waker(&mut self, receiver_id: u32, waker: &Waker) {
+        self.receivers
+            .get_mut(&receiver_id)
+            .unwrap()
+            .update_waker(waker);
+    }
+}
+
+// /////////////////////////////////////////////////////////////////////////////
+//
+// SENDER SIDE
+//
+// /////////////////////////////////////////////////////////////////////////////
+
+pub struct Sender<T>
+where
+    T: Clone,
+{
+    id: u32,
+    pending: AtomicUsize,
+    inner: Arc<Mutex<Inner<T>>>,
+}
+
+#[allow(clippy::missing_panics_doc)]
+impl<T> Sender<T>
+where
+    T: Clone,
+{
+    pub fn subscribe(&self) -> Result<Receiver<T>, SubscribeError> {
+        let id = {
+            let mut lock = self.inner.lock().unwrap();
+            lock.create_receiver(false)
+        }?;
+
+        Ok(Receiver {
+            id,
+            pending: Default::default(),
+            inner: self.inner.clone(),
+        })
+    }
+
+    pub fn try_send(&self, v: T) -> Result<(), SendError<T>> {
+        let mut lock = self.inner.lock().unwrap();
+        lock.try_send(v)
+    }
+}
+
+#[allow(clippy::missing_panics_doc)]
+impl<T> Sender<T>
+where
+    T: Clone + Unpin,
+{
+    pub fn send(&self, v: T) -> Send<T> {
+        let _ignore = self.pending.fetch_add(1, Ordering::Relaxed);
```

Comment:
Do you have deeper knowledge of `Ordering`s? For myself, I often guess :shrug:. Is this ordering chosen with determination or just because it seems to work?